core/src/main/java/tech/pegasys/eth2signer/core/multikey/MultiKeyArtifactSignerProvider.java

```diff
@@ -12,72 +12,140 @@
  */
 package tech.pegasys.eth2signer.core.multikey;
 
-import tech.pegasys.eth2signer.core.multikey.metadata.SigningMetadataFile;
-import tech.pegasys.eth2signer.core.multikey.metadata.UnencryptedKeyMetadataFile;
-import tech.pegasys.eth2signer.core.signers.unencryptedfile.UnencryptedKeyFileSignerFactory;
+import tech.pegasys.eth2signer.core.multikey.metadata.SignerParser;
 import tech.pegasys.eth2signer.core.signing.ArtifactSigner;
 import tech.pegasys.eth2signer.core.signing.ArtifactSignerProvider;
 
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.DirectoryStream.Filter;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.stream.Collectors;
 
+import org.apache.commons.io.FilenameUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class MultiKeyArtifactSignerProvider implements ArtifactSignerProvider, MultiSignerFactory {
+public class MultiKeyArtifactSignerProvider implements ArtifactSignerProvider {
 
   private static final Logger LOG = LogManager.getLogger();
+  private final Path configsDirectory;
+  private SignerParser signerParser;
 
-  private final SigningMetadataTomlConfigLoader signingMetadataTomlConfigLoader;
-
-  public MultiKeyArtifactSignerProvider(
-      final SigningMetadataTomlConfigLoader signingMetadataTomlConfigLoader) {
-    this.signingMetadataTomlConfigLoader = signingMetadataTomlConfigLoader;
+  public MultiKeyArtifactSignerProvider(final Path rootDirectory, final SignerParser signerParser) {
+    this.configsDirectory = rootDirectory;
+    this.signerParser = signerParser;
   }
 
   @Override
   public Optional<ArtifactSigner> getSigner(final String signerIdentifier) {
-    return signingMetadataTomlConfigLoader
-        .loadMetadataForAddress(signerIdentifier)
-        .map(metadataFile -> metadataFile.createSigner(this));
+    final String normalisedIdentifier = normaliseIdentifier(signerIdentifier);
+    final Optional<ArtifactSignerWithFileName> signer =
+        loadSignerForIdentifier(normalisedIdentifier);
+    if (signer.isEmpty()) {
+      LOG.error("No matching metadata file found for the identifier {}", signerIdentifier);
+      return Optional.empty();
+    }
+    final ArtifactSignerWithFileName signerWithFileName = signer.get();
+    if (!signerMatchesIdentifier(signerWithFileName, signerIdentifier)) {
```

Comment:
nit: I suspect this is the check which is wrong - the real verification is "does loaded Signer have the same identifier as that requested?"
The concept of "filenames" is irrelevant here.

Reply:
It's not actually checking the filename, just poorly named variable. I've made a tuple with the signer and filename used internally so I can display details of the failure to the user.